File: src/data_loader.cpp
#include "data_loader.h"
#include <fstream>
#include <sstream>
#include <iostream>
#include <vector>
#include <string>

// function will read CSV and convert rows into StockRecord objects
std::vector<StockRecord> loadDataset(const std::string& filename) {
    std::vector<StockRecord> records;
    std::ifstream file(filename);
    std::string line;

    if (!file.is_open()) {
        std::cerr << "Error: cannot open dataset: " << filename << std::endl;
        return records;
    }

    // read/discard header row
    std::getline(file, line);

    while (std::getline(file, line)) {
        if (line.empty()) {
            continue;
        }

        std::stringstream lineStream(line);
        std::string columnValue;
        std::vector<std::string> columns;

        while (std::getline(lineStream, columnValue, ',')) {
            columns.push_back(columnValue);
        }

        if (columns.size() != 9) {
            continue;
        }

        try {
            StockRecord record;

            // assign values to fields in StockRecord
            record.date = columns[0];
            record.open = std::stod(columns[1]);
            record.high = std::stod(columns[2]);
            record.low = std::stod(columns[3]);
            record.close = std::stod(columns[4]);
            record.adjClose = std::stod(columns[5]);
            record.volume = std::stoll(columns[6]);
            record.ticker = columns[7];
            record.company = columns[8];

            records.push_back(record);
        }
        catch (...) {
            continue;
        }
    }

    return records;
}

File: data_loader.py
from dataclasses import dataclass
from typing import List


# rows from dataset
# records should filter by ticker
# analyze by whatever metric user picks for top-k
# use full StockRecord objects for output
# compare records using the selected metric
@dataclass
class StockRecord:
    date: str
    open: float
    high: float
    low: float
    close: float
    adjClose: float
    volume: int
    ticker: str
    company: str


# function will read CSV and convert rows into StockRecord objects
def loadDataset(filename: str) -> List[StockRecord]:
    records: List[StockRecord] = []

    try:
        with open(filename, "r", encoding="utf-8") as file:
            # read/discard header row
            file.readline()

            for line in file:
                line = line.strip()

                if line == "":
                    continue

                columns = line.split(",")

                # expected columns: 0. date, 1. open, 2. high, 3. low, 4. close,
                # 5. adjClose, 6. volume, 7. ticker, 8. company
                if len(columns) != 9:
                    continue

                try:
                    # assign values to fields in StockRecord
                    record = StockRecord(
                        date=columns[0],
                        open=float(columns[1]),
                        high=float(columns[2]),
                        low=float(columns[3]),
                        close=float(columns[4]),
                        adjClose=float(columns[5]),
                        volume=int(columns[6]),
                        ticker=columns[7],
                        company=columns[8]
                    )

                    records.append(record)

                except ValueError:
                    continue

    except FileNotFoundError:
        print("Error: cannot open dataset", filename)

    return records